### ISSUE TITLE
Passed all test cases but failed CodeCrunch because of Floating Point Precision Error

### ISSUE THREAD

**Issue by hingen** [nus-cs2030/2122-s1#177]
Not an issue, just a heads up that can hopefully help someone.

So my code passed all provided test cases. But when I uploaded it to CodeCrunch, it failed a test case. Spent 2 hours debugging only to realise it was caused by a floating point precision error.

I set a threshold of 1e-15 which is fine for all the provided test cases. However, CodeCrunch has a test case that uses over 600 customers. For my code, the threshold of 1e-15 was too small and this caused it to produce incorrect behaviour. Changing the threshold fixed the issue.

**Comment by BrendanCheong** [comment]
Hmmm interesting, I used a precision of 1e-9 and it was fine for CodeCrunch.

**Comment by hingen** [comment]
> Hmmm interesting, I used a precision of 1e-9 and it was fine for CodeCrunch.

Yes, 1e-9 seems sufficient. For me, changing to 1e-10 was enough to pass CodeCrunch.

**Comment by RussellDash332** [comment]
You used precision? 😕

**Comment by bingsen0806** [comment]
I used 1e-15 too but no problem tho...

**Comment by hingen** [comment]
> I used 1e-15 too but no problem tho...

oh, dat's interesting. I guess it depends on how the code is written. Perhaps my code does more mutable operations on double variables which increases the chance of such an error occurring. The appropriate threshold likely varies depending on how the code is written and how much precision is needed.

**Comment by thojianxiong** [comment]
Can I check which part did y'all use precision threshold for? Didn't seem to need to include it when I was working on the Project!

**Comment by hingen** [comment]
> Can I check which part did y'all use precision threshold for? Didn't seem to need to include it when I was working on the Project!

I sorted all my events by timings and use the threshold to check if two events are suppose to execute at the same time. If they are, then I execute the one with the lowest customer ID. The precision error caused my events to execute out of order.

**Comment by KeithPJX** [comment]
Just completed the project, I did not seem to need this too. I guess it's just because of the difference in the direction we took to implement the simulator.

**Comment by Hongyi6328** [comment]
I have done the project. I don't think there is a need to use precision. Instead, you can try non-strict inequality to test something. I did it and it was fine
